### PR TITLE
Support anydata for client data binding

### PR DESCRIPTION
**Pull request by chamil321**
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec

### CODE REVIEW

File: ballerina/http_types.bal [hunk 1]
```diff
@@ -30,11 +30,11 @@ public type Service distinct service object {
 
 };
 
-# The types of the response payload that are returned by the HTTP `client` after the data binding operation.
-public type PayloadType string|xml|json|map<string>|map<json>|byte[]|record {| anydata...; |}|record {| anydata...;|}[];
+//# The types of the response payload that are returned by the HTTP `client` after the data binding operation.
+//public type PayloadType string|xml|json|map<string>|map<json>|byte[]|record {| anydata...; |}|record {| anydata...;|}[];
```
Comment (TharmiganK) [inline ballerina/http_types.bal:34]: We can remove these commented lines right?

File: native/src/main/java/io/ballerina/stdlib/http/api/client/actions/AbstractHTTPAction.java [hunk 2]
```diff
@@ -551,4 +551,17 @@ private void addHttpStatusCode(int statusCode) {
...
+//        HttpHeaders headers = (HttpHeaders) response.getNativeData(HTTP_HEADERS);
+//        String contentType = headers.get(HttpHeaderNames.CONTENT_TYPE.toString());
+//
+//        AbstractPayloadBuilder payloadBuilder = getBuilder(contentType, typedesc.getDescribingType());
+//        try {
+//            return payloadBuilder.getValue(entity, readonly);
+//        } catch (BError error) {
+//            return error;
+//        }
+//    }
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/client/actions/AbstractHTTPAction.java:566]: Do we need this commented method?

File: native/src/main/java/io/ballerina/stdlib/http/api/service/signature/builder/AbstractPayloadBuilder.java [hunk 3]
```diff
@@ -44,18 +42,15 @@ public abstract class AbstractPayloadBuilder {
     private static final String URL_ENCODED_PATTERN = "^.*x-www-form-urlencoded.*$";
 
     /**
-     * Build the inbound payload, bind it to the respective type and return the next signature param index.
+     * Get the built the inbound payload after binding it to the respective type.
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/service/signature/builder/AbstractPayloadBuilder.java:45]:
```suggestion
     * Get the built inbound payload after binding it to the respective type.
```

File: docs/spec/spec.md [hunk 4]
```diff
@@ -526,13 +526,58 @@ See section [Query] to understand accessing query param via the request object.
...
 
-The payload parameter eases access of request payload during the resource invocation. When the payload param is 
-defined with @http:Payload annotation, the listener binds the inbound request payload into the param type and return. 
-The type of payload param can be one of the anytype.
-
-The payload binding process begins soon after the finding the correct resource for the given URL and before the 
-resource execution. Based on the `Content-type` header, the payload it deserialized. If the header is not present or 
-not a standard header, the deserializer is inferred by the param type.
+The payload parameter is used to access the request payload during the resource invocation. When the payload param is 
+defined with @http:Payload annotation, the listener deserialize the inbound request payload based on the mime type
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:530]:
```suggestion
defined with @http:Payload annotation, the listener deserialize the inbound request payload based on the media type 
```

File: docs/spec/spec.md [hunk 5]
```diff
@@ -526,13 +526,58 @@ See section [Query] to understand accessing query param via the request object.
...
-defined with @http:Payload annotation, the listener binds the inbound request payload into the param type and return. 
-The type of payload param can be one of the anytype.
-
-The payload binding process begins soon after the finding the correct resource for the given URL and before the 
-resource execution. Based on the `Content-type` header, the payload it deserialized. If the header is not present or 
-not a standard header, the deserializer is inferred by the param type.
+The payload parameter is used to access the request payload during the resource invocation. When the payload param is 
+defined with @http:Payload annotation, the listener deserialize the inbound request payload based on the mime type 
+which retrieved by the `Content-type` header of the request. The data binding happens thereafter considering the 
+parameter type. The type of payload param can be one of the `anytype`. If the header is not present or not a
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:532]:
```suggestion
parameter type. The type of payload parameter can be one of the `anytype`. If the header is not present or not a 
```

File: docs/spec/spec.md (hunk 4 above)
Comment (shafreenAnfar) [reply docs/spec/spec.md:530]: It seems media type is the correct terms which is formally known as MIME type.

File: docs/spec/spec.md [hunk 6]
```diff
@@ -526,13 +526,58 @@ See section [Query] to understand accessing query param via the request object.
...
-The type of payload param can be one of the anytype.
-
-The payload binding process begins soon after the finding the correct resource for the given URL and before the 
-resource execution. Based on the `Content-type` header, the payload it deserialized. If the header is not present or 
-not a standard header, the deserializer is inferred by the param type.
+The payload parameter is used to access the request payload during the resource invocation. When the payload param is 
+defined with @http:Payload annotation, the listener deserialize the inbound request payload based on the mime type 
+which retrieved by the `Content-type` header of the request. The data binding happens thereafter considering the 
+parameter type. The type of payload param can be one of the `anytype`. If the header is not present or not a 
+standard header, the binding type is inferred by the param type.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:533]:
```suggestion
standard header, the binding type is inferred by the parameter type.
```

File: docs/spec/spec.md [hunk 7]
```diff
@@ -526,13 +526,58 @@ See section [Query] to understand accessing query param via the request object.
...
+
+Following table explains the compatible `anydata` types with each common mime type. In the absence of a standard mime
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:535]:
```suggestion
Following table explains the compatible `anydata` types with each common media type. In the absence of a standard mime 
```

File: docs/spec/spec.md [hunk 8]
```diff
@@ -526,13 +526,58 @@ See section [Query] to understand accessing query param via the request object.
...
-resource execution. Based on the `Content-type` header, the payload it deserialized. If the header is not present or 
-not a standard header, the deserializer is inferred by the param type.
+The payload parameter is used to access the request payload during the resource invocation. When the payload param is 
+defined with @http:Payload annotation, the listener deserialize the inbound request payload based on the mime type 
+which retrieved by the `Content-type` header of the request. The data binding happens thereafter considering the 
+parameter type. The type of payload param can be one of the `anytype`. If the header is not present or not a 
+standard header, the binding type is inferred by the param type.
+
+Following table explains the compatible `anydata` types with each common mime type. In the absence of a standard mime 
+type, the binding type is inferred by the payload param type itself. If the type is not compatible with the mime type,
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:536]:
```suggestion
type, the binding type is inferred by the payload param type itself. If the type is not compatible with the media type,
```

File: docs/spec/spec.md [hunk 9]
```diff
@@ -1187,8 +1232,52 @@ The HTTP client remote function supports the contextually expected return types.
 infer the expected payload type from the LHS variable type. This is called as client payload binding support where the 
 inbound response payload is accessed and parse to the expected type in the method signature. It is easy to access the
 payload directly rather manipulation `http:Response` using its support methods such as `getTextPayload()`, ..etc.
-Client data binding supports `anydata` where the payload is deserialized based on the content type before binding it 
-to the required type.
+
+Client data binding supports `anydata` where the payload is deserialized based on the mime type before binding it
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1236]:
```suggestion
Client data binding supports `anydata` where the payload is deserialized based on the media type before binding it 
```

File: docs/spec/spec.md [hunk 10]
```diff
@@ -1187,8 +1232,52 @@ The HTTP client remote function supports the contextually expected return types.
 infer the expected payload type from the LHS variable type. This is called as client payload binding support where the 
 inbound response payload is accessed and parse to the expected type in the method signature. It is easy to access the
 payload directly rather manipulation `http:Response` using its support methods such as `getTextPayload()`, ..etc.
-Client data binding supports `anydata` where the payload is deserialized based on the content type before binding it 
-to the required type.
+
+Client data binding supports `anydata` where the payload is deserialized based on the mime type before binding it 
+to the required type. Similar to the service data binding following table explains the compatible `anydata` types with 
+each common mime type. In the absence of a standard mime type, the binding type is inferred by the payload param type
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1238]:
```suggestion
each common mime type. In the absence of a standard media type, the binding type is inferred by the payload param type 
```

Comment (shafreenAnfar) [inline docs/spec/spec.md:1238]:
```suggestion
each common mime type. In the absence of a standard media type, the binding type is inferred by the payload parameter type 
```

File: docs/spec/spec.md [hunk 11]
```diff
@@ -1187,8 +1232,52 @@ The HTTP client remote function supports the contextually expected return types.
 infer the expected payload type from the LHS variable type. This is called as client payload binding support where the 
 inbound response payload is accessed and parse to the expected type in the method signature. It is easy to access the
 payload directly rather manipulation `http:Response` using its support methods such as `getTextPayload()`, ..etc.
-Client data binding supports `anydata` where the payload is deserialized based on the content type before binding it 
-to the required type.
+
+Client data binding supports `anydata` where the payload is deserialized based on the mime type before binding it 
+to the required type. Similar to the service data binding following table explains the compatible `anydata` types with 
+each common mime type. In the absence of a standard mime type, the binding type is inferred by the payload param type 
+itself. If the type is not compatible with the mime type, error is returned.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1239]:
```suggestion
itself. If the type is not compatible with the media type, error is returned.
```

**Review by TharmiganK — APPROVED** [review]

**Pull request merged**